### PR TITLE
bug: add instances for empty rows

### PR DESCRIPTION
**Pull request by emanuelaepure10**
Add the correct number of instances also for the cases when the rows contains empty cell as result of a formula 

ING-4135

### CODE REVIEW

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java [hunk 1]
```diff
@@ -110,20 +111,23 @@ public static Workbook loadWorkbook(InputStream input, URI location, boolean xls
...
-			int count = 0;
-			for (int i = header.getFirstCellNum(); i < header.getLastCellNum(); i++) {
-				Cell cell = header.getCell(i);
-				String text = extractText(cell, sheet);
-				// cell cannot be empty to extract the text
-				if (text != null) {
-					headerCell(count, text);
-					count++;
+	protected void analyseHeader(Sheet sheet, int skipNlines) {
+		if (skipNlines <= 0) {
```
[the remark this answers is not in the document]
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:117]: No effect. I can move the check when the method is called.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java [hunk 2]
```diff
@@ -140,13 +144,15 @@ protected void analyseHeader(Sheet sheet) {
 	 * 
 	 * @param sheet the table sheet
 	 */
-	private void analyseContent(Sheet sheet) {
+	private void analyseContent(Sheet sheet, int skipNlines) {
+		int startingLine = 0;
```
[the remark this answers is not in the document]
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:148]: My understanding was that if you have 0 as skip lines then the first line is the header. 
but it's very true that you might have already the first instance in the first line. 

In the case of having an header this one would be added twice: once to the header and once to the rows, so you would add 2 instances for the same header. 
but in case you have no header, you need to consider also the first line in the content.

Why we don't add a checkbox in case there is a header and make the life easier to the user to understand and also to us?

[the remark this answers is not in the document]
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:150]: If 0 lines should be skipped then the first line/row already contains data and no header.

For one thing we can't just change the behavior of this setting and thus create backwards compatibility issues.
If we feel there is the need to improve the user experience that should be part of a separate task.

[the remark this answers is not in the document]
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java]: Done. now works with all the possible cases.
Thank you for your patience :-)

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java [hunk 3]
```diff
@@ -71,7 +72,9 @@ protected void analyse(LocatableInputSupplier<? extends InputStream> source, boo
 			evaluator = wb.getCreationHelper().createFormulaEvaluator();
 
 			// the first row represents the header
-			analyseHeader(sheet);
+			if (skipNlines <= 0) {
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:75]: Negate the condition?

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java [hunk 4]
```diff
@@ -136,17 +139,15 @@ protected void analyseHeader(Sheet sheet) {
...
 
 	/**
-	 * Analyse the table content.
+	 * Analyse the table content
 	 * 
 	 * @param sheet the table sheet
 	 */
 	private void analyseContent(Sheet sheet) {
 		// for each row starting from the second
 		for (int i = 1; i <= sheet.getLastRowNum(); i++) {
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:148]: Shouldn't the start be based on the `skipNLines` setting?
The current behavior (skipping one row) I would only expect for `skipNLines == 1`.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java (hunk 3 above)
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:75]: If no lines are to be skipped that means there is no header to be analysed because the first row is already instance data.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java [hunk 5]
```diff
@@ -136,17 +139,21 @@ protected void analyseHeader(Sheet sheet) {
...
 	 * 
 	 * @param sheet the table sheet
+	 * @param skipNlines number of lines to skip
 	 */
-	private void analyseContent(Sheet sheet) {
-		// for each row starting from the second
-		for (int i = 1; i <= sheet.getLastRowNum(); i++) {
+	private void analyseContent(Sheet sheet, int skipNlines) {
+		int startLine = skipNlines;
+		if (skipNlines <= 0) {
```
[the remark this answers is not in the document]
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:151]: Because the first line has already been analyse in the header if skipNlines <= 0.
And if the header is a header or just a row with data the first line will be added to the instances anyhow.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java (hunk 3 above)
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:75]: If we want to skip one line, then we assume that first row is the header, but in that case `analyseHeader` is not called at all anymore.
The case with multiple lines to skip is harder to assess, that was obviously not in the original concept of this implementation. In line with what was done before would be to analyse the first line in this case as well.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:151]: If that's the argument then in my view `analyseHeader` should not only be called when `skipNLines` is zero or less (see other comment).

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java (hunk 3 above)
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java]: For me makes no sense to analyse the header if anyway is not going to be used for the case with skipNlines > 0.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:149]: > the first line will be added to the instances anyhow

I think I found the [code](https://github.com/halestudio/hale/blob/3de05597fc72a1b6c7b0639e2b6bba50374088e6/io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/reader/XLSInstanceReader.java#L176) you are referring to.

But wouldn't it make more sense if you introduce to the API in `AbstractAnalyseTable` that lines should be skipped, implying that the skipped lines are not part of the data, that in case there are no line to be skipped and thus no dedicated header, that it would report all rows as data rows?
In my view the instance reader should just not add the header then. This was done previously because `AbstractAnalyseTable` so far represented the concept of a table where the first row is the header and the rest are data.

[the remark this answers is not in the document]
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:149]: Also in my view the header should not be added.
Doesn't make any sense to have the header as a data as long as doesn't contain any data.

[the remark this answers is not in the document]
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:149]: So do we then agree that the following is true? 

- the instance reader should not add the header provided by `AbstractAnalyseTable` as instance
- `AbstractAnalyseTable` should call `analyseRow` for the first row if `skipNLines` is less or equal to zero

[the remark this answers is not in the document]
Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:149]: @stempler yes. and the code at the moment is exactly in line with what you have mention.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/reader/XLSInstanceReader.java [hunk 6]
```diff
@@ -165,24 +169,18 @@ private void loadSheet(SheetInfo sheet, IOReporter reporter) throws Exception {
...
 			while (allRows.hasNext()) {
 				List<String> row = allRows.next();
-				addInstanceForRow(row, type, propAr, line, reporter);
-
+				if (row != null && !row.stream().allMatch(s -> s == null || s.isEmpty())) {
+					addInstanceForRow(row, type, propAr, line, reporter);
+				}
 				line++;
 			}
 		}
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/reader/XLSInstanceReader.java:186]: Related to my comment [here](https://github.com/halestudio/hale/pull/1087/files#r1425178823), the first point would imply removing this whole section.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (stempler) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:149]: > and the code at the moment is exactly in line with what you have mention.

I'm afraid we disagree there.

- My first point would mean removing the whole section on `if (skipNlines <= 0)` from the instance reader (see new comment).
- My second point would mean not setting `startLine` to `1` here, so the data row is part of the rows that are analysed with `analyseRow`

**Comment by emanuelaepure10** [comment]
In summary, the 'skipNlines' is employed to skip a specified number of lines that do not contain data. 

However, when dealing with `headers` intended for import as instances, one need to pay caution regarding the placement of 'skipNlines'.

File: io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java [hunk 7]
```diff
@@ -136,17 +137,16 @@ protected void analyseHeader(Sheet sheet) {
...
+	 * Analyse the table content if skipNlines <=0 that don't analyse first row,
+	 * which has already been analyse into the header else analyse starting with
+	 * the skip line
 	 * 
 	 * @param sheet the table sheet
 	 */
 	private void analyseContent(Sheet sheet) {
-		// for each row starting from the second
-		for (int i = 1; i <= sheet.getLastRowNum(); i++) {
+		for (int i = 0; i <= sheet.getLastRowNum(); i++) {
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java:147]: Why did you change this to start at `0` and not at `skipNlines`? I thought that was the whole point of the changes to `AbstractAnalyseTable`?

Comment (emanuelaepure10) [reply io/plugins/eu.esdihumboldt.hale.io.xls/src/eu/esdihumboldt/hale/io/xls/AbstractAnalyseTable.java]: I feel that we are spinning around the tail.
Perhaps that commit aligns with your thoughts.

**Review by stempler — APPROVED** [review]
